CATE validation suite

amarvenu opened the pull request:
Add new CATE validation class, DRTester, to perform doubly-robust based CATE validation methods. Implemented methods include BLP, calibration, and QINI coefficient. Relevant files are:

- econml/validation/drtester.py  (main class)
- econml/tests/test_drtester.py  (tests)
- notebooks/CATE validation.ipynb  (example usage notebook)

Tests cover 100% of implemented code (see attached image).

![drtester_coverage](https://github.com/py-why/EconML/assets/89877778/558d9245-0ff4-485d-9bab-b589fd77d6eb)

Looking forward to any comments.

vsyrgkanis commented on econml/validate/drtester.py:
```diff
@@ -0,0 +1,796 @@
+
+    def __init__(
+        self,
+        reg_outcome,
```
For consistency with the DRLearner API, let's call these `model_regression` and `model_propensity`. Just so that we don't introduce yet another naming convention in the package.

vsyrgkanis commented on econml/validate/drtester.py:
```diff
@@ -0,0 +1,796 @@
+    Direct Marketing Analytics Journal (2007), pages 14–21.
+    """
+
+    def __init__(
```
All arguments in __init__ of classes should be keyword only arguments. So add `,*,` after the `self`, to force keyword arguments.

vsyrgkanis commented on econml/validate/drtester.py:
```diff
@@ -0,0 +1,796 @@
+        self,
+        reg_outcome,
+        reg_t,
+        n_splits: int = 5
```
Let's make this `cv` and following the exact same concentions on what `cv` can be as in DRLearner. So cv can be a number or a list of splits etc. This is a scikit-learn naming convention.

vsyrgkanis commented on econml/validate/drtester.py:
```diff
@@ -0,0 +1,796 @@
+        self.reg_outcome = reg_outcome
+        self.reg_t = reg_t
+        self.n_splits = n_splits
+        self.dr_train = None
```
You should not be initializing variables inside `__init__`. This is a sklearn convention and for a reason that relates to cloning conventions. Maybe initialize this inside fit. Also better give them names that end with `_` to denote an internal variable that is a stateful property produced by fit. e.g. `self.cate_preds_train_`, `self.cate_preds_val_` etc.

vsyrgkanis commented on econml/validate/drtester.py:
```diff
@@ -0,0 +1,796 @@
+
+        return self
+
+    def fit_cate(
```
I am confused by this method. The evaluation class should not be fitting a cate estimator. It should only be evaluating an already fitted CATE estimator.

vsyrgkanis commented on econml/validate/drtester.py:
```diff
@@ -0,0 +1,796 @@
+
+        Parameters
+        ----------
+        reg_cate: estimator, default ``None``
```
The cate estimator should be adhering to the EconML CATE API. This is not fit and predict. Rather it is fit and effect(T=1), effect(T=2), const_marginal_effect etc. 

Also here you should not be using `fit` you should be assuming the CATE estimator is already fitted. So just specify the methods that are required to be supported by the fitted CATE estimator. I suspect this would just be `cate.effect(T=t)`, where t are the different treatment values sorted in lexicographic order.

vsyrgkanis replied:
Make sure you can handle the case when the input treatments are for instance alphanumeric, i.e. I pass a T of the form ['a', 'b', 'a', 'a', 'b', etc]

vsyrgkanis commented on econml/validate/drtester.py:
```diff
@@ -0,0 +1,796 @@
+            df_plot1['tmt'] = self.tmts[k + 1]
+            self.df_plot = pd.concat((self.df_plot, df_plot1))
+
+        return self
```
Most prob this should be returning some evaluation result e.g. the calibration score

vsyrgkanis replied:
Maybe you should create a class CalibrationEvaluationResults that contains all these variables you describe and has the API for parsing it, e.g. the plot_cal, most prob should be part of this new class of `CalibrationEvaluationResults`

vsyrgkanis commented on econml/validate/drtester.py:
```diff
@@ -0,0 +1,796 @@
+
+        Parameters
+        ----------
+        reg_cate: estimator, default ``None''
```
Similar comment here

vsyrgkanis commented on econml/validate/drtester.py:
```diff
@@ -0,0 +1,796 @@
+            {'treatment': self.tmts[1:], 'blp_est': params, 'blp_se': errs, 'blp_pval': pvals}
+        ).round(3)
+
+        return self
```
Similarly maybe here this should be returning a instance a class `BLPEvaluationResults`

vsyrgkanis commented on econml/validate/drtester.py:
```diff
@@ -0,0 +1,796 @@
+            {'treatment': self.tmts[1:], 'qini_coeff': qinis, 'qini_se': errs, 'qini_pval': pvals},
+        ).round(3)
+
+        return self
```
This most prob should return QiniEvalutationResults object

vsyrgkanis commented on econml/validate/drtester.py:
```diff
@@ -0,0 +1,796 @@
+        self.df_res = self.blp_res.merge(self.qini_res, on='treatment')
+        self.df_res['cal_r_squared'] = np.around(self.cal_r_squared, 3)
+
+        return self
```
Maybe here you can return an EvaluationResults object that has each of them as a property, i.e. 
res = evaluate_all(cate)
res.blp.summary()
res.cal.summary()
res.qini.summary()
res.cal.plot_cal(...)
res.summary() # plots all summary results all together

vsyrgkanis commented on econml/validate/drtester.py:
```diff
@@ -0,0 +1,796 @@
+
+    # CV prediction of CATEs
+    @staticmethod
+    def fit_cate_cv(
```
I suspect all these methods related to CATE fitting should not be here. you should only be accepting fitted cate models.

vsyrgkanis reviewed (commented):
Great job! Very well documented and clean code. Several comments and requests, but overall seams in a great shape!

amarvenu commented:
Updated code to address the given comments. Branch underlying PR has been updated. Modified test cases still have 100% code coverage:
![cate_coverage_080123](https://github.com/py-why/EconML/assets/89877778/c60082d0-afcc-4119-9fee-668e5319d0be)

kbattocchi reviewed (approved):
Sorry for the terrible delay in approving this; it will be a great addition to the library and it looks good to me, I've just made a few small changes to clean up some linting issues and such.
